File: internal/aws/cognito_auth.go
package aws

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider/types"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

type CognitoAuth struct {
	client       *cognitoidentityprovider.Client
	userPoolId   string
	clientId     string
	clientSecret string
	region       string
}

func NewCognitoAuth(cfg *Config) (*CognitoAuth, error) {
	if !cfg.CognitoEnabled {
		return nil, fmt.Errorf("Cognito is not enabled")
	}
	
	if cfg.CognitoUserPoolId == "" || cfg.CognitoClientId == "" {
		return nil, fmt.Errorf("Cognito user pool ID and client ID are required")
	}
	
	ctx := context.Background()
	
	var awsCfg aws.Config
	var err error
	
	if cfg.CognitoAccessKey != "" && cfg.CognitoSecretKey != "" {
		awsCfg, err = config.LoadDefaultConfig(ctx,
			config.WithRegion(cfg.CognitoRegion),
			config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
				cfg.CognitoAccessKey,
				cfg.CognitoSecretKey,
				"",
			)),
		)
	} else {
		awsCfg, err = config.LoadDefaultConfig(ctx,
			config.WithRegion(cfg.CognitoRegion),
		)
	}
	
	if err != nil {
		return nil, fmt.Errorf("failed to load AWS config: %w", err)
	}
	
	cognitoClient := cognitoidentityprovider.NewFromConfig(awsCfg)
	
	return &CognitoAuth{
		client:       cognitoClient,
		userPoolId:   cfg.CognitoUserPoolId,
		clientId:     cfg.CognitoClientId,
		clientSecret: cfg.CognitoClientSecret,
		region:       cfg.CognitoRegion,
	}, nil
}

func (ca *CognitoAuth) AuthenticateUser(username, password string) (*types.AuthenticationResultType, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	
	authParams := map[string]string{
		"USERNAME": username,
		"PASSWORD": password,
	}
	
	if ca.clientSecret != "" {
		authParams["SECRET_HASH"] = ca.clientSecret
	}
	
	input := &cognitoidentityprovider.InitiateAuthInput{
		AuthFlow:       types.AuthFlowTypeUserPasswordAuth,
		ClientId:       aws.String(ca.clientId),
		AuthParameters: authParams,
	}
	
	result, err := ca.client.InitiateAuth(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("failed to authenticate with Cognito: %w", err)
	}
	
	return result.AuthenticationResult, nil
}

func (ca *CognitoAuth) GetUser(accessToken string) (*cognitoidentityprovider.GetUserOutput, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	
	input := &cognitoidentityprovider.GetUserInput{
		AccessToken: aws.String(accessToken),
	}
	
	result, err := ca.client.GetUser(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("failed to get user from Cognito: %w", err)
	}
	
	return result, nil
}

func (ca *CognitoAuth) VerifyToken(token string) (bool, error) {
	_, err := ca.GetUser(token)
	if err != nil {
		return false, err
	}
	
	return true, nil
}

func SetupCognitoAuth(app *pocketbase.PocketBase, cognito *CognitoAuth) error {
	app.OnRecordAuthRequest("users").BindFunc(func(e *core.RecordAuthRequestEvent) error {
		authResult, err := cognito.AuthenticateUser(e.Identity, e.Password)
		if err != nil {
			log.Printf("Cognito authentication failed, falling back to PocketBase: %v", err)
			return nil
		}
		
		log.Printf("User authenticated via Cognito: %s", e.Identity)
		
		record, err := app.FindFirstRecordByData("users", "email", e.Identity)
		if err != nil {
			collection, err := app.FindCollectionByNameOrId("users")
			if err != nil {
				return fmt.Errorf("failed to find users collection: %w", err)
			}
			
			record = core.NewRecord(collection)
			record.Set("email", e.Identity)
			record.Set("emailVisibility", true)
			record.Set("verified", true)
			
			if err := app.Save(record); err != nil {
				return fmt.Errorf("failed to create user record: %w", err)
			}
			
			log.Printf("Created new user from Cognito: %s", e.Identity)
		}
		
		e.Record = record
		
		if authResult.AccessToken != nil {
			record.Set("cognitoAccessToken", *authResult.AccessToken)
		}
		if authResult.RefreshToken != nil {
			record.Set("cognitoRefreshToken", *authResult.RefreshToken)
		}
		if authResult.IdToken != nil {
			record.Set("cognitoIdToken", *authResult.IdToken)
		}
		
		return nil
	})
	
	log.Println("Cognito authentication hooks installed")
	return nil
}

func (ca *CognitoAuth) ListUsers(limit int32) ([]types.UserType, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	
	input := &cognitoidentityprovider.ListUsersInput{
		UserPoolId: aws.String(ca.userPoolId),
		Limit:      aws.Int32(limit),
	}
	
	result, err := ca.client.ListUsers(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("failed to list users from Cognito: %w", err)
	}
	
	return result.Users, nil
}

func (ca *CognitoAuth) CreateUser(email, password string, attributes map[string]string) error {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	
	userAttrs := []types.AttributeType{
		{
			Name:  aws.String("email"),
			Value: aws.String(email),
		},
		{
			Name:  aws.String("email_verified"),
			Value: aws.String("true"),
		},
	}
	
	for name, value := range attributes {
		userAttrs = append(userAttrs, types.AttributeType{
			Name:  aws.String(name),
			Value: aws.String(value),
		})
	}
	
	input := &cognitoidentityprovider.AdminCreateUserInput{
		UserPoolId:        aws.String(ca.userPoolId),
		Username:          aws.String(email),
		UserAttributes:    userAttrs,
		TemporaryPassword: aws.String(password),
		MessageAction:     types.MessageActionTypeSuppress, // Don't send welcome email
	}
	
	_, err := ca.client.AdminCreateUser(ctx, input)
	if err != nil {
		return fmt.Errorf("failed to create user in Cognito: %w", err)
	}
	
	setPwdInput := &cognitoidentityprovider.AdminSetUserPasswordInput{
		UserPoolId: aws.String(ca.userPoolId),
		Username:   aws.String(email),
		Password:   aws.String(password),
		Permanent:  true,
	}
	
	_, err = ca.client.AdminSetUserPassword(ctx, setPwdInput)
	if err != nil {
		return fmt.Errorf("failed to set user password: %w", err)
	}
	
	return nil
}

File: pb_hooks/aws/cognito_auth.js

const { 
  CognitoIdentityProviderClient, 
  InitiateAuthCommand, 
  GetUserCommand,
  ListUsersCommand,
  AdminCreateUserCommand,
  AdminSetUserPasswordCommand
} = require('@aws-sdk/client-cognito-identity-provider');

class CognitoAuth {
  constructor(config) {
    if (!config.cognitoEnabled) {
      throw new Error('Cognito is not enabled');
    }
    
    if (!config.cognitoUserPoolId || !config.cognitoClientId) {
      throw new Error('Cognito user pool ID and client ID are required');
    }
    
    const clientConfig = {
      region: config.cognitoRegion,
    };
    
    if (config.cognitoAccessKey && config.cognitoSecretKey) {
      clientConfig.credentials = {
        accessKeyId: config.cognitoAccessKey,
        secretAccessKey: config.cognitoSecretKey,
      };
    }
    
    this.client = new CognitoIdentityProviderClient(clientConfig);
    this.userPoolId = config.cognitoUserPoolId;
    this.clientId = config.cognitoClientId;
    this.clientSecret = config.cognitoClientSecret;
    this.region = config.cognitoRegion;
  }
  
  async authenticateUser(username, password) {
    const authParams = {
      USERNAME: username,
      PASSWORD: password,
    };
    
    if (this.clientSecret) {
      authParams.SECRET_HASH = this.clientSecret;
    }
    
    const command = new InitiateAuthCommand({
      AuthFlow: 'USER_PASSWORD_AUTH',
      ClientId: this.clientId,
      AuthParameters: authParams,
    });
    
    const result = await this.client.send(command);
    return result.AuthenticationResult;
  }
  
  async getUser(accessToken) {
    const command = new GetUserCommand({
      AccessToken: accessToken,
    });
    
    const result = await this.client.send(command);
    return result;
  }
  
  async verifyToken(token) {
    try {
      await this.getUser(token);
      return true;
    } catch (err) {
      return false;
    }
  }
  
  async listUsers(limit = 60) {
    const command = new ListUsersCommand({
      UserPoolId: this.userPoolId,
      Limit: limit,
    });
    
    const result = await this.client.send(command);
    return result.Users;
  }
  
  async createUser(email, password, attributes = {}) {
    const userAttributes = [
      {
        Name: 'email',
        Value: email,
      },
      {
        Name: 'email_verified',
        Value: 'true',
      },
    ];
    
    for (const [name, value] of Object.entries(attributes)) {
      userAttributes.push({
        Name: name,
        Value: value,
      });
    }
    
    const createCommand = new AdminCreateUserCommand({
      UserPoolId: this.userPoolId,
      Username: email,
      UserAttributes: userAttributes,
      TemporaryPassword: password,
      MessageAction: 'SUPPRESS',
    });
    
    await this.client.send(createCommand);
    
    const setPasswordCommand = new AdminSetUserPasswordCommand({
      UserPoolId: this.userPoolId,
      Username: email,
      Password: password,
      Permanent: true,
    });
    
    await this.client.send(setPasswordCommand);
  }
}

function setupCognitoAuth(app, cognito) {
  app.onRecordAuthWithPasswordRequest('users').add(async (e) => {
    try {
      const authResult = await cognito.authenticateUser(e.identity, e.password);
      
      console.log(`User authenticated via Cognito: ${e.identity}`);
      
      let record;
      try {
        record = await app.findFirstRecordByData('users', 'email', e.identity);
      } catch (err) {
        const collection = await app.findCollectionByNameOrId('users');
        record = new Record(collection);
        record.set('email', e.identity);
        record.set('emailVisibility', true);
        record.set('verified', true);
        
        await app.save(record);
        
        console.log(`Created new user from Cognito: ${e.identity}`);
      }
      
      e.record = record;
      
      if (authResult.AccessToken) {
        record.set('cognitoAccessToken', authResult.AccessToken);
      }
      if (authResult.RefreshToken) {
        record.set('cognitoRefreshToken', authResult.RefreshToken);
      }
      if (authResult.IdToken) {
        record.set('cognitoIdToken', authResult.IdToken);
      }
      
    } catch (err) {
      console.log(`Cognito authentication failed, falling back to PocketBase: ${err.message}`);
      return e.next();
    }
  });
  
  console.log('Cognito authentication hooks installed');
}

module.exports = {
  CognitoAuth,
  setupCognitoAuth
};
